templates/cluster-template.yaml

```diff
@@ -189,7 +195,9 @@ spec:
           kubeletExtraArgs:
             cloud-provider: external
           name: '{{ ds.meta_data["local_hostname"] }}'
-      preKubeadmCommands: []
+      preKubeadmCommands:
+      - echo '10.0.0.100   ${CLUSTER_NAME}-${APISERVER_LB_DNS_SUFFIX}.${AZURE_LOCATION}.cloudapp.azure.com'
```

Comment:
It might be cleaner to do this in `postKubeadmCommands`:

```
- echo '10.0.0.100 $(grep 'server:' /etc/kubernetes/kubelet.conf | awk -F[/:] '{print $5}') >> /etc/hosts
```

This approach would allow us to entirely skip the need to pass in the dnsName manually in another part of the template, and we'd just rely upon the existing foo and kubeadm to tell us this info.

I'm pretty sure we'd need to do this as a `postKubeadmCommand`, and not `preKubeadmCommand`, kubeadm command. I would assume that kubeadm hasn't yet paved the kubeconfig by the time we invoke `preKubeadmCommands`.

Reply:
> This approach would allow us to entirely skip the need to pass in the dnsName manually in another part of the template, and we'd just rely upon the existing foo and kubeadm to tell us this info.

I like the idea. One less variable to manage.

However, isn't it better to setup custom DNS resolution for the API Server in `preKubeadmCommands` than in `postKubeadmCommands`? 
By using `preKubeadmCommand`, in this scenario, we can be sure that no Azure network policies will interfere in the `kubeadm join`step. 
Meaning, if Azure policies (blocking internet access on the VM) were to kick in immediately after a VM comes up, custom DNS resolution for the API Server to the internal IP should help `kubeadm join`.  

What do you say?

Reply:
Will `/etc/kubernetes/kubelet.conf` be present at the time that `preKubeadmCommands` runs? I assume that kubeadm is responsible for paving that file, which means we have to wait.

But that may not be practical for Windows nodes that come online later (it may be too late and as you mention, kubeadm join may fail and block `postKubeadmCommands` from ever executing.

cc @jsturtevant

Reply:
> Will /etc/kubernetes/kubelet.conf be present at the time that preKubeadmCommands runs?

I don't think it will be present, kubeadm creates it.